finality-aleph/src/validator_network/mod.rs

```diff
@@ -0,0 +1,69 @@
+#![allow(dead_code)]
+use std::fmt::Display;
+
+use aleph_primitives::AuthorityId;
+use codec::Codec;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+mod io;
+#[cfg(test)]
+mod mock;
+
+/// What the data sent using the network has to satisfy.
+pub trait Data: Clone + Codec + Send + Sync + 'static {}
+
+impl<D: Clone + Codec + Send + Sync + 'static> Data for D {}
+
+/// Network represents an interface for opening and closing connections with other Validators,
+/// and sending direct messages between them.
+///
+/// Note on Network reliability and security: it is neither assumed that the sent messages must be
+/// always delivered, nor the established connections must be secure in any way. The Network
+/// implementation might fail to deliver any specific message, so messages have to be resend while
+/// they still should be delivered.
+#[async_trait::async_trait]
+pub trait Network<A: Data, D: Data>: Send {
+    /// Add the peer to the set of connected peers.
+    fn add_connection(&mut self, peer: AuthorityId, addresses: Vec<A>);
+
+    /// Remove the peer from the set of connected peers and close the connection.
+    fn remove_connection(&mut self, peer: AuthorityId);
+
+    /// Send a message to a single peer.
+    /// This function should be implemented in a non-blocking manner.
+    fn send(&self, data: D, recipient: AuthorityId);
+
+    /// Receive a message from the network.
+    async fn next(&mut self) -> Option<D>;
```

Comment:
Why doesn't it output the sender as well?

Reply:
We don't need it, if it turns out we actually need it for something in the future we can add it then.

Reply:
Yeah, we don't need it because the API of the AlephBFT network is suboptimal (does not require the sender). And it is suboptimal because we initially assumed the information about the sender will be hard to get. It is actually easy to get, and here we are throwing it away for free.

What I'm proposing is to break the cycle, and actually start improving the network by making the right choices here.

Reply:
Ok fine. The only remaining thing I'm worried about currently is that this will hinder the ability to log suitably precise diagnostic information. Feel free to resolve if you think it should not be a problem.

Reply:
Logging will mostly happen on a different level which will have access to peer information anyway.